api/rest/coordinator.go

```diff
@@ -0,0 +1,271 @@
+package rest
+
+import (
+	"archive/zip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const numberOfWorkers = 10
+const contextDoneErrMsg = "bundle creation context finished before bundle creation finished"
+
+// BundleStatus tracks the status of local bundle creation requests
+type bundleStatus struct {
+	id   string
+	node node
+	done bool
+	err  error
+}
+
+// golangcli-lint marks this as dead code because nothing uses the interface
+// yet. this will be removed with the bundle handler PR
+//nolint:deadcode
+// coordinator is an interface to coordinate the creation of diagnostics bundles
+// across a cluster of nodes
+type coordinator interface {
+	// CreateBundle starts the bundle creation process. Status updates be monitored
+	// on the returned channel.
+	CreateBundle(ctx context.Context, id string, nodes []node) <-chan bundleStatus
+	// CollectBundle waits until all the nodes' bundles have finished, downloads,
+	// and merges them. The resulting bundle zip file path is returned.
+	CollectBundle(ctx context.Context, bundleID string, numBundles int, statuses <-chan bundleStatus) (string, error)
+}
+
+// ParallelCoordinator implements Coordinator interface to coordinate bundle
+// creation across a cluster, parallelized across multiple goroutines.
+type parallelCoordinator struct {
+	client Client
+
+	// statusCheckInterval defines how often the status of the local bundles will
+	// be checked
+	statusCheckInterval time.Duration
+	workDir             string
+}
+
+// job is a function that will be called by the worker function. The output will be added to results channel
+type job func() bundleStatus
+
+// worker is a function that will run incoming jobs from jobs channel and put jobs output to results chan
+func worker(ctx context.Context, jobs <-chan job, results chan<- bundleStatus) {
+	for {
+		select {
+		case j := <-jobs:
```

Comment:
```suggestion
		for j :=  range jobs {
```

And we need to find a way to close jobs channel

Reply:
How about having information about number of expected number of bundles and count finished bundles and exit when no more job need to be done?